keps/173-headless-service-per-replica/README.md

```diff
@@ -0,0 +1,359 @@
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+```golang
+type LeaderWorkerSetSpec struct {
+ // SubdomainPolicy determines the policy that will be used when creating
+ // the headless service
+ SubdomainPolicy SubdomainPolicy `json:"subdomainPolicy,omitempty"`
+}
+
+type NetworkConfig struct {
+  SubdomainPolicy SubdomainPolicy
+}
+type SubdomainPolicy string
+const (
+  // SubdomainShared will create a single headless service that all replicas 
+  // will share. The host names look like:
+  // Replica 0: my-lws-0.my-lws, my-lws-0-1.my-lws
+  // Replica 1: my-lws-1.my-lws, my-lws-1-1.my-lws
+  SubdomainShared SubdomainPolicy = "Shared"
+  // SubdomainUniquePerReplica will create a headless service for each
+  // leader-worker group. 
+  // The leader host names will look like:
+  // Replica 0: my-lws-0.my-lws
+  // Replica 1: my-lws-1.my-lws
+  // The worker host names will look like:
+  // Replica 0: my-lws-0-1.my-lws-0, my-lws-0-2.my-lws-0
+  // Replica 1: my-lws-1-1.my-lws-1, my-lws-1-2.my-lws-1
+  SubdomainUniquePerReplica SubdomainPolicy = "UniquePerReplica"
+)
+```
+
+### Implementation
+
+With SubdomainPolicy set to SubdomainUniquePerReplica, we will create a headless
+service for all of the leaders, and a headless service per replica for the workers. In order 
+to ensure backwards compatability, the default value if non is set will be Shared.
+
+A label will be added to determine whether a pod is a leader or a worker.
```

Comment:
We can use `workerIndex = 0` to select the leaders, but cannot do `workerIndex != 0` to select the workers since AFAIK selector does not have that capability.